RFCs/0009-Fn::ForEach.md

```diff
@@ -0,0 +1,1461 @@
+
+```yaml
+AWSTemplateFormatVersion: 2010-09-09
+Transform: 'AWS::LanguageExtensions'
+Resources:
+  SNSTopic:
+    Type: AWS::SNS::Topic
+  'Fn::ForEach::Topics':
+    - !Ref SNSTopic
+    - ['A', 'B', 'C']
+    - 'SNSTopic${Identifier}':
+        Type: AWS::SNS::Topic
+```
+
+### 7. The “Collection” and each value in the “Collection” must be resolvable before Resource creation
+
+* All the values in the list defined within the 2nd parameter have to be *known* before resource provisioning, otherwise the Stack/ChangeSet creation/update. You can’t, for example, refer to an attribute of a Resource. Values must be known before CloudFormation performs any remote resource interactions.
+* The template below is invalid because an element in the “Collection” cannot be resolved before Resource creation.
+
+```yaml
+AWSTemplateFormatVersion: 2010-09-09
+Transform: 'AWS::LanguageExtensions'
+Resources:
+  SNSTopic:
+    Type: AWS::SNS::Topic
+  'Fn::ForEach::Topics':
+    - Identifier
+    - ['A', 'B', !Ref SNSTopic]
+    - 'SNSTopic${Identifier}':
+        Type: AWS::SNS::Topic
+```
+
+* The template below is invalid because the “Collection” cannot be resolved before Resource creation.
+
+```yaml
+AWSTemplateFormatVersion: 2010-09-09
+Transform: 'AWS::LanguageExtensions'
+Resources:
+  VPC:
+    Type: 'AWS::EC2::VPC'
+    Properties:
+      CidrBlock: 10.0.0.0/16
+      EnableDnsSupport: 'true'
+      EnableDnsHostnames: 'true'
+  TargetGroup:
+    Type: AWS::ElasticLoadBalancingV2::TargetGroup
+    Properties:
+      Name: 'MyElasticGroup'
+      Port: 80
+      Protocol: 'HTTP'
+      VpcId: !Ref VPC
+  'Fn::ForEach::Topics':
+    - Identifier
+    - !GetAtt [TargetGroup, LoadBalancerArns]
+    - 'SNSTopic${Identifier}':
+        Type: AWS::SNS::Topic
+```
+
+## Potential Follow-up Features
+
+Features like supporting iterating over a key-valur pair or allowing Ref/Fn::GetAtt on the UniqueLoopName are out of scope for this RFC.
+
+The GitHub issue below contains potential follow-up features, where based on customer demand, CloudFormation will create separate RFC’s for those customer pain-points:
+
+https://github.com/aws-cloudformation/cfn-language-discussion/issues/118
+
+
+## Appendix
+
+### Appendix A: Pros and Cons matrix for looping function names
+
+|FunctionName	|Pros	|Cons	|
+|---	|---	|---	|
+|Fn::ForEach	|Addresses valid feedback in GitHub pull requestIntrinsic function name indicates it’s a functional concept and not an imperative oneForEach Illustrates that we’re operating on a template fragment over a set of valuesGives customers familiar with coding concepts the idea that some form of “looping” (or replication) is done	|Intrinsic function name is potentially not unclear to customers who aren’t familiar with coding concepts; however, the syntax of Fn::ForEach::<ResourceLogicalId> would give users an idea that within the function, we define the template fragment "for each" resource	|
```

Comment:
The formatting on these table entries is a mess

Reply:
Nice catch! Looks like the resolved markdown looks ok, but I've updated the raw markdown to be clean as well.